pkg/controller/appsync/appsync.go

```diff
@@ -12,40 +12,152 @@
-	db     *database.Database
-	h      *render.Renderer
-}
-
-// New creates a new appsync controller.
-func New(config *config.AppSyncConfig, db *database.Database, h *render.Renderer) (*Controller, error) {
-	return &Controller{
-		config: config,
-		db:     db,
-		h:      h,
-	}, nil
-}
+const playStoreHost = `play.google.com/store/apps/details`
 
 // HandleSync performs the logic to sync mobile apps.
-func (c *Controller) HandleSync(ctx context.Context) http.Handler {
+func (c *Controller) HandleSync() http.Handler {
+	type AppSyncResult struct {
+		OK     bool    `json:"ok"`
+		Errors []error `json:"errors,omitempty"`
+	}
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		// TODO(whaught): implement this
-		controller.InternalError(w, r, c.h, errors.New("not implemented"))
+		ctx := r.Context()
+		apps, err := clients.AppSync(c.config.AppSyncURL, c.config.Timeout, c.config.FileSizeLimitBytes)
+		if err != nil {
+			controller.InternalError(w, r, c.h, err)
+			return
+		}
+
+		// If there are any errors, return them
+		if merr := c.syncApps(ctx, apps); merr != nil {
+			if errs := merr.WrappedErrors(); len(errs) > 0 {
+				c.h.RenderJSON(w, http.StatusInternalServerError, &AppSyncResult{
+					OK:     false,
+					Errors: errs,
+				})
+				return
+			}
+		}
+		c.h.RenderJSON(w, http.StatusOK, &AppSyncResult{OK: true})
 	})
 }
+
+// syncApps looks up the realm and associated list of MobileApps for each entry of AppsResponse. Then it
+// checks to see if there exists an app with the AppResponse SHA hash, if not it creates a new MobileApp.
+func (c *Controller) syncApps(ctx context.Context, apps *clients.AppsResponse) *multierror.Error {
+	logger := logging.FromContext(ctx).Named("appsync.syncApps")
+	var merr *multierror.Error
+
+	realms := map[string]*database.Realm{}
+	appsByRealm := map[uint][]*database.MobileApp{}
+
+	for _, app := range apps.Apps {
+
+		realm, err := c.findRealmForApp(app, realms)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("unable to lookup realm for region %q: %w", app.Region, err))
+			continue
+		}
+
+		realmApps, err := c.findAppsForRealm(realm.ID, appsByRealm)
+		if err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("unable to list apps for realm %d: %w", realm.ID, err))
+			continue
+		}
+
+		// Find out if this realm's applist already has an app with this fingerprint.
+		hasSHA, hasGeneratedName := false, false
+		for _, a := range realmApps {
+			if a.SHA == app.SHA256CertFingerprints {
+				hasSHA = true
+			}
+			if a.Name == generateAppName(app) {
+				hasGeneratedName = true
+			}
+		}
+
+		// Didn't find an app. make one.
+		if !hasSHA {
+			logger.Infow("app not found during sync, adding", "app", app)
+
+			name := generateAppName(app)
+			if hasGeneratedName { // add a random string to names on collision
+				s, err := project.RandomString()
+				if err != nil {
+					logger.Errorw("error generating app name", "error", err)
```

Comment:
Any reason not to add to the multierror instead?

Reply:
nah let's do that. more consistent